Development/nmos/ocsp_utils.cpp

```diff
@@ -0,0 +1,194 @@
+                }
+
+                return ocsp_req_;
+            }
+
+#if !defined(_WIN32) || !defined(__cplusplus_winrt) || defined(CPPREST_FORCE_HTTP_CLIENT_ASIO)
+            // this callback is called when client includes a certificate status request extension in the TLS handshake
+            int server_certificate_status_request(SSL* ssl, void* arg)
+            {
+                auto& ocsp_state = *(nmos::experimental::ocsp_state*)arg;
+
+                return nmos::experimental::set_ocsp_response(ssl, nmos::with_read_lock(ocsp_state.mutex, [&] { return ocsp_state.ocsp_response; })) ? SSL_TLSEXT_ERR_OK : SSL_TLSEXT_ERR_NOACK;
+            }
+#endif
+        }
+
+        // get a list of OCSP URIs from server certificate
+        std::vector<web::uri> get_ocsp_uris(const std::string& certificate)
+        {
+            using ssl::experimental::BIO_ptr;
+            using ssl::experimental::X509_ptr;
+
+            BIO_ptr bio(BIO_new(BIO_s_mem()), &BIO_free);
+            if ((size_t)BIO_write(bio.get(), certificate.data(), (int)certificate.size()) != certificate.size())
+            {
+                throw ocsp_exception("failed to get_ocsp_uris while loading server certificate to BIO: BIO_new failure: " + ssl::experimental::last_openssl_error());
+            }
+            X509_ptr x509(PEM_read_bio_X509_AUX(bio.get(), NULL, NULL, NULL), &X509_free);
+            if (!x509)
+            {
+                throw ocsp_exception("failed to get_ocsp_uris while converting server certificate BIO to X509: PEM_read_bio_X509_AUX failure: " + ssl::experimental::last_openssl_error());
+            }
+            auto ocsp_uris_ = X509_get1_ocsp(x509.get());
+
+            std::vector<web::uri> ocsp_uris;
+            for (int idx = 0; idx < sk_OPENSSL_STRING_num(ocsp_uris_); idx++)
+            {
+                web::uri ocsp_uri(utility::s2us(sk_OPENSSL_STRING_value(ocsp_uris_, idx)));
+                if (ocsp_uris.end() == std::find_if(ocsp_uris.begin(), ocsp_uris.end(), [&](const web::uri& uri) { return uri == ocsp_uri; }))
+                {
+                    ocsp_uris.push_back(ocsp_uri);
+                }
+            }
+            return ocsp_uris;
+        }
+
+        // construct an OCSP request from the specified list of server certificate chains
+        std::vector<uint8_t> make_ocsp_request(const std::vector<std::string>& certificate_chains)
+        {
+            if (certificate_chains.empty())
+            {
+                throw ocsp_exception("failed to make_ocsp_request: no server certificate chains");
+            }
+
+            // a minimal server certificate chain starts with the server's certificate, followed by the server's issuer certificate.
+            // for now, assuming all the server certificates are issued by the same issuer, let's get the issuer certificate from the 1st server certificate chain.
```

Comment:
I never thought that there could be a difference between TLS 1.2. and TLS 1.3, for the OCSP request, do you have the RFC in hand?

Reply:
I am assuming both certificates are issued from the same issuer. I will need to study how to handle OCSP for the different issuers.

ca -> immediate ca 1 ->certificate 1
ca -> immediate ca 2 ->certificate 2

Reply:
Regarding TLS 1 3, RFC 8446 says:

> It also changes how Online Certificate Status Protocol (OCSP) messages are carried and therefore updates [[RFC6066](https://datatracker.ietf.org/doc/html/rfc6066)] and obsoletes [[RFC6961](https://datatracker.ietf.org/doc/html/rfc6961)] as
   described in [Section 4.4.2.1](https://datatracker.ietf.org/doc/html/rfc8446#section-4.4.2.1).

That obsoleted RFC 6961 is the spec for supporting multiple stapled certs in TLS 1.2. Since it sounds like you've already tested that, maybe OpenSSL is implementing that for you? Probably it implements the TLS 1.3 way too. We just need tests 😁

Reply:
At the moment in nmos-cpp settings, we can only allow for 1 root CA.

Reply:
I don't think that's true... we allow for 1 ca certificate _file_, but it may contain multiple authorities?

Reply:
What am I thinking, the Root CA is for verifying the other server's certificate, which we are making calls to. Ignore my previous message.

Reply:
Ah, that's true. The server certs are likely to be rooted at (one of) the same CAs, 
but maybe that's beside the point like you say.

Also... I had forgotten that although we can be configured with multiple server certs with different issuers, actually we can only send one of those chains to a client... which one should be selected by OpenSSL based on what algorithms the client says it supports, etc. See e.g. https://tools.ietf.org/html/rfc8446#section-4.4.2.2

So that leads to interesting question... don't we need to know which one, to decide which one to fetch and return the OCSP response for?! Is there some way that OpenSSL tells us?!